src/HttpCache/TagCollector.php

```diff
@@ -0,0 +1,40 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\HttpCache;
+
+use ApiPlatform\Metadata\ApiProperty;
+use ApiPlatform\Serializer\TagCollectorInterface;
+use Symfony\Component\PropertyInfo\Type;
+
+/**
+ * Collects cache tags during normalization.
+ *
+ * @author Urban Suppiger <urban@suppiger.net>
+ */
+class TagCollector implements TagCollectorInterface
+{
+    public function collect(mixed $object = null, string $format = null, array $context = [], string $iri = null, mixed $data = null, string $attribute = null, ApiProperty $propertyMetadata = null, Type $type = null): void
```

Comment:
I think we're closed to something nice, a few things still bothers me. Do we really need `data` and `object` ? I think that this works using only `object`. 
If I think about an IriCollector for cache tags, I don't see the format in its signature, we can put this inside our $context no? I also think that `$propertyMetadata` alone is fine as it should hold the `$type` information. So my signature looks like this: 

```
    public function collect(mixed $object = null, string $iri = null, array $context = [], string $attribute = null, ApiProperty $propertyMetadata = null): void
```

Last but not least, I also think that the `iri` can actually be computed by the tag collector if needed, so we'd end up with something like this (`$context['iri']` should be defined for performances if available): 


```
    public function collect(mixed $object = null, array $context = [], string $attribute = null, ApiProperty $propertyMetadata = null): void
```

Not even sure if we couldn't remove `$attribute` and `$propertyMetadata`, are these really useful? 

You could probably add your use case as functional test so that I see how you use it? You can add your service in our `config_common.yml` inside the TestBundle.

Reply:
> You could probably add your use case as functional test so that I see how you use it? You can add your service in our `config_common.yml` inside the TestBundle.

I tried to integrate this into our own project yesterday. That might help a bit to evaluate some of the use cases:
https://github.com/ecamp/ecamp3/pull/3610/files#diff-92df153f2ab4331ccfd55e92699e78c72a3223ff10a38cb870ad59f769308fe0

Can also try to integrate some of this into the test suit.

For your other comments, will look at them and respond later.

Reply:
I added a functional test with commit https://github.com/api-platform/core/pull/5758/commits/a404dd7fd2b6d05057ecfea1c7b79687182edd0a

Adding the custom TagCollector to `config_common.yml` obviously breaks the existing tests in `features/http_cache/tags.feature`. Not sure what the best way is to isolate this test and couldn't find a good example to copy from. Do I need to set up a separate behat suite which only adds the tag collector service?

Reply:
https://github.com/api-platform/core/pull/5758#discussion_r1307370466 should resolve the tests issue no?